### PR TITLE
[quant][graphmode][fx] Add support for weight and bias dtype in backend_config_dict

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68229
* #68228
* #68227
* #68042
* #67537
* __->__ #68602

Summary:
This PR adds support for configuring weight/bias dtype in backend_config_dict
and refactor the current code that checks when to insert observers

Test Plan:
Regression
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```
Actual test of functionality will be in next PR
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D32537712](https://our.internmc.facebook.com/intern/diff/D32537712)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).

File: torch/ao/quantization/fx/prepare.py [hunk 1]
```diff
@@ -372,6 +367,27 @@ def get_arg_target_dtype(
     else:
         return node_name_to_target_dtype[arg.name]["output_activation_dtype"]
 
+def get_arg_target_dtype_as_input(
```
Comment (vkuzo) [inline torch/ao/quantization/fx/prepare.py:370]: sorry, what does this mean conceptually?  Is `get_arg_target_dtype_as_output` getting the output dtype of arg, and `get_arg_target_dtype_as_input` getting the input dtype of arg?  Or something else?

Comment (vkuzo) [reply torch/ao/quantization/fx/prepare.py:370]: after reading the code I get it...how about something like
`get_target_dtype_of_arg_from_previous_node` and `get_target_dtype_of_arg_in_cur_node`?

Comment (vkuzo) [reply torch/ao/quantization/fx/prepare.py:370]: or maybe `get_arg_target_dtype_as_output` and `get_arg_target_dtype_as_input_to_this_node`?

**Review by vkuzo — APPROVED** [review]

File: torch/ao/quantization/fx/prepare.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/quantization/fx/prepare.py:370]: yeah I feel this is a bit hard to name as well, I will use something like `get_arg_target_dtype_as_output` and `get_arg_target_dtype_as_input_to_this_node`

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32537712).

**Pull request closed without merging**